### ISSUE TITLE
propagate onClose from original Stream

### ISSUE THREAD

**Issue by shollander** [javasync/streamemo#2]
This is a really neat utility that fills a hole in java 8 Streams.

One thing that would be nice to have is that ability to close the original stream. Many Streams represent a resource that needs to be closed once the stream is finished (e.g. [Files.list()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Files.html#list(java.nio.file.Path))) It would be nice if the `onClose()` method of a stream from the `Supplier.get()` would trigger a call to the original Stream's `onClose()` method. This would ensure that all resources are closed when they are no longer needed.

**Comment by fmcarvalho** [comment]
It makes sense. Would you like to make a Pull Request with it?

**Comment by shollander** [comment]
I tried to add a simple unit test, but for some reason I get a strange error when running the test. I get the following stacktrace:
```
java.lang.IllegalStateException: Accept exceeded fixed size of 0
	at java.base/java.util.stream.Nodes$FixedNodeBuilder.accept(Nodes.java:1232)
	at org.javasync.streams.Replayer$Recorder.lambda$0(Replayer.java:93)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:302)
	at org.javasync.streams.Replayer$Recorder.getOrAdvance(Replayer.java:91)
	at org.javasync.streams.Replayer$Recorder$MemoizeIter.tryAdvance(Replayer.java:110)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:326)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:550)
	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:517)
	at org.javasync.streams.test.ReplayTest.testOnClose(ReplayTest.java:163)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:436)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:170)
	at org.junit.jupiter.engine.execution.ThrowableCollector.execute(ThrowableCollector.java:40)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:166)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:113)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:58)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:112)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$2(HierarchicalTestExecutor.java:120)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.lambda$executeRecursively$3(HierarchicalTestExecutor.java:120)
	at org.junit.platform.engine.support.hierarchical.SingleTestExecutor.executeSafely(SingleTestExecutor.java:66)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.executeRecursively(HierarchicalTestExecutor.java:108)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor$NodeExecutor.execute(HierarchicalTestExecutor.java:79)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:55)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:43)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:229)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:197)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:211)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:191)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:137)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
```

Do you have ay idea what is causing this error? I don't think it has anything to do with the `onClose()` that I added.

**Comment by fmcarvalho** [comment]
@shollander I am not figuring out the reason for that behavior. I noticed that changing your unit test with the following workaround works fine: 

```java
try(Stream<Long> s = replayer.get()) {
   List<Long> lst = s.collect(toList());
   Long[] numbers = lst.toArray(new Long[lst.size()]);
   assertArrayEquals(new Long[] {1L, 2L, 3L, 4L}, numbers);
}
```

There is a strange behavior on `toArray()` of the `Stream`. It is wrapping a consumer in a `FixedNodeBuilder` .

**Comment by fmcarvalho** [comment]
Thankful to your unit test we are observing a bug from here: 
```java
public MemoizeIter(){
  super(estimateSize, getSrcIter().characteristics());
}
```

I am not understanding why. Stopping in debug here `super(estimateSize, getSrcIter().characteristics()); ` we can watch that `estimateSize` has value 4. But then jumping to super `AbstractSpliterator` constructor it receives `0` as argument!!!

**Comment by fmcarvalho** [comment]
Apparently an optimization is gathering the wrong value for `estimateSize` field.  Since the `getSrcIter()` lazily fetch the source stream the initial value of `estimateSize` is 0 and only then is updated with the right estimated size after the invocation of `getSrcIter()`.

This bug fix implies changing the `MemoizeIter` constructor to:

```
public MemoizeIter(Spliterator<T> srcIter){
  super(srcIter.estimateSize(), srcIter.characteristics());
}
```
and also its call to `new MemoizeIter(getSrcIter())`. 

After this fix, now I have another unit test failing, the `testHighlightDifferentAspects`.

**Comment by fmcarvalho** [comment]
@shollander The fix that I propose in my previous comment solves the streamemo Bug observed in your unit test. With that fix you may see your unit test passing.

Yet, your unit test is no fully evaluating the `onClose` propagating feature.  If the stream source came from a supplier that freshly creates a new stream then your implementation will propagate the `onClose`  to the wrong stream. I changed your unit test proposal with the following code to show you a scenario where it fails:

```java
    	final List<AtomicInteger> closeCounter = new ArrayList<>();
    	final Supplier<Stream<Long>> originalStream = () -> {
            Stream<Long> src = LongStream.range(1, 5).boxed();
            AtomicInteger c = new AtomicInteger();
            closeCounter.add(c);
            src.onClose(() -> c.getAndIncrement());
            return src;
    	};

    	Supplier<Stream<Long>> replayer = Replayer.replay(originalStream);
    	try(Stream<Long> s = replayer.get()) {
    		Long[] numbers = s.toArray(Long[]::new);
    		assertArrayEquals(new Long[] {1L, 2L, 3L, 4L}, numbers);
    	}
        closeCounter.forEach(c -> assertEquals(1, c.get()));
```

At the end of this test the `closeCounter` list will have 2 entries and you are not incrementing the right one. In truth it should only contain one entry. So it will also violate the following assertion: `assertEquals(1, closeCounter.size());`

Thank you

**Comment by shollander** [comment]
@fmcarvalho I believe I have fixed all this issues.

**Comment by fmcarvalho** [comment]
@shollander I have already merged your contribution and made a new release 1.0.1 including the onClose propagation feature.

Please give a star to our repository streamemo to atract the community attention.

Thank you

**Comment by shollander** [comment]
@fmcarvalho Great, thanks!
